### PR TITLE
bump to @bentley/react-scripts@4.0.3, eslint@^7.11.0

### PR DESCRIPTION
**Pull request by aruniverse**
Bumping @bentley/react-scripts to 4.0.3, see https://github.com/imodeljs/create-react-app/pull/35

Moved some of the flags from the package.json scripts into an `.env` file, users can override it with a `.env.*.local` file  with different setting for development or production builds in a `.env.development` or `.env.production` file

The core monorepo is using `react@16.14.0` which supports the [new react-jsx transform](https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html), but we couldn't use it until typescript was bumped to 4.1 or we bumped react-scripts to 4.0.0. I have explicitly disabled the use of this new transform, but we can re-enable later. Currently there is an issue using that feature because we pin down our `@types/react` down to 16.9.43 and its missing the type for the transform. "Could not find a declaration file for module 'react/jsx-runtime'"

Eitherway, with this change, everything still compiles. Due to outstanding issues with the test-apps w.r.t breaking changes in 2.13.x i was unable to test it out thoroughly, but everything should be fine.

### CODE REVIEW

**Comment by pmconne** [comment]
> Due to outstanding issues with the test-apps w.r.t breaking changes in 2.13.x i was unable to test it out thoroughly, but everything should be fine.

Can you elaborate? Are the test-apps broken in some way?

File: test-apps/display-test-app/package.json
Comment (calebmshafer) [inline test-apps/display-test-app/package.json:77]: Did you need to remove `eslint` here for this to work?

File: test-apps/ui-test-app/package.json
Comment (calebmshafer) [inline test-apps/ui-test-app/package.json:65]: This should still be used by the backend code, even though the frontend has it bundled.

**Review by calebmshafer — CHANGES_REQUESTED** [review]
@aruniverse before we merge this, we'll want to make sure to merge the CRA 4 PR in our fork and get an official version published.  I'll review that later today.

**Comment by aruniverse** [comment]
> > Due to outstanding issues with the test-apps w.r.t breaking changes in 2.13.x i was unable to test it out thoroughly, but everything should be fine.
> 
> Can you elaborate? Are the test-apps broken in some way?

I believe there's an issue with auth in ui-test-app that @ramanujam-raman is actively working on.

Paul i did _test_ the display-test-app, i was able to open a local model, and everything seemed to work, but not too familiar with the app to test it thoroughly.

File: test-apps/display-test-app/package.json
Comment (aruniverse) [reply test-apps/display-test-app/package.json:77]: I removed it because its a direct dep of react-scripts and should be included, but youre right its used by the backend so i should explicitly still depend on it. Wasn't sure if i should also bump eslint throughout the monorepo to ^7.11.0.

I did have to set `SKIP_PREFLIGHT_CHECK=true` in the `.env ` file because during compilation it was complaining about the two different versions of eslint. For the frontend this is fine and can be ignored because we are now using the `DISABLE_ESLINT_PLUGIN` flag

**Review by pmconne — APPROVED** [review]
Approving display-test-app changes...can't comment on CRA stuff.

File: test-apps/display-test-app/package.json
Comment (calebmshafer) [reply test-apps/display-test-app/package.json:77]: Ah okay, yes you can go ahead and update the ESLint version.  This update to CRA 4 was actually what we were waiting on before updating the repo.

File: extensions/map-layers/package.json
Comment (DanEastBentley) [inline extensions/map-layers/package.json:47]: Does this version of react-scripts include the fix in this PR for the updated version of immer: https://github.com/facebook/create-react-app/issues/10411

Comment (aruniverse) [reply extensions/map-layers/package.json:47]: it does :)

File: test-apps/ui-test-app/package.json
Comment (aruniverse) [reply test-apps/ui-test-app/package.json:65]: re-added, nice catch. thx

File: test-apps/display-test-app/package.json
Comment (aruniverse) [reply test-apps/display-test-app/package.json:77]: bumped to ^7.11.0

File: test-apps/display-performance-test-app/.env
Comment (calebmshafer) [inline test-apps/display-performance-test-app/.env:1]: Can this be removed now that we updated `eslint`?

File: test-apps/ui-test-app/.env
Comment (calebmshafer) [inline test-apps/ui-test-app/.env:1]: @aruniverse we'll have to see how this plays with the `rush copy:config` script.  Any reason to not keep the `.env.template`? The intent is to have someone copy it and create their own `.env`.

Comment (calebmshafer) [reply test-apps/ui-test-app/.env:1]: Also surprised you were able to even commit it :)

**Review by DanEastBentley — APPROVED** [review]

File: test-apps/display-performance-test-app/.env
Comment (aruniverse) [reply test-apps/display-performance-test-app/.env:1]: I tried that, but still ran into issues. For some reason the global eslint ver in the monorepo is still 6.8.0. Not sure why the lock file didnt get updated to 7.21.0. I didn't want to manually play with the lock file, and since we're disabling eslint i thought it was fine. Any idea how to clean this up in rush?

image

File: test-apps/ui-test-app/.env
Comment (aruniverse) [reply test-apps/ui-test-app/.env:1]: Was unaware of the internal-tools copy-config script and what it does. 
imo, we should provide the `.env` file, and users will override it with their values for their imodel and etc in a `.env.local`. I can add comments to the `.env` file alerting the user to do this.

I was able to commit these changes because i updated the `.gitignore` 😅

**Review by grigasp — APPROVED** [review]

File: test-apps/display-performance-test-app/.env
Comment (calebmshafer) [reply test-apps/display-performance-test-app/.env:1]: Strange... I'll pull and see what's going on.

File: docs/changehistory/2.6.0.md
Comment (calebmshafer) [inline docs/changehistory/2.6.0.md:22]: Can you revert this change?  2.6.0 won't support 7.x :)

File: docs/changehistory/2.5.0.md
Comment (calebmshafer) [inline docs/changehistory/2.5.0.md:130]: Here too

File: test-apps/display-test-app/package.json
Comment (calebmshafer) [inline test-apps/display-test-app/package.json:77]: Did you mean to delete this line?

File: docs/changehistory/2.6.0.md
Comment (aruniverse) [reply docs/changehistory/2.6.0.md:22]: Great catch! Bad find all and replace...
I'll come back and resolve these when our official 4.0 ver is out

**Comment by aruniverse** [comment]
I'm going to create an issue with the CopyWebpackPlugins repo once i get back next week, but heres my findings so far as to why the new CopyStaticAssetsPlugin isn't working with pnpm workspaces, and part of a draft of what that issue will look like. (just to refresh my memory next week, posting here in advance and also because i need more time to draft the issue and im out of time for the day :( )

----------------------------------------------------------------------------------------------------------

My rush based monorepo just switched to using pnpm workspaces, and the CopyWebpackPlugin didn't work due to dependencies being hoisted, and it's only partially resolved by using `require.resolve`. 

Previously my config looked like:

```ts
new CopyPlugin({
  patterns: [
    {
      from: '**/public/**/*',
      context: 'node_modules/@MY_SCOPE',
      noErrorOnMissing: true,
      to({ absoluteFilename }: { absoluteFilename: string }) {
        const regex = new RegExp('(public(?:\\\\|\/))(.*)');
        return regex.exec(absoluteFilename)![2];
      },
    },
    ...
  ],
}),
```
Basically I have a few (15-20) packages (some with different scopes) that deliver some assets I need for my app. I would previously look in node_modules and try to copy over everything in those package's public dir into my apps public dir.

With the hoisting now its unable to correctly find those dependencies and the assets never get copied over.

The `require.resolve` does work and I'm able to resolve the path to where the dependency gets hoisted too, but now I need to manually specify all the deps to look for, where as previously I just needed to specify the scope. This is a minor inconvenience and i have a workaround like so:

```ts
const appPackageJson = require(paths.appPackageJson);
const pkgsICareAbout = Object.keys(appPackageJson.dependencies).filter(n => n.startsWith('@scope_1') || n.startsWith('@scope_2'));
const copyPatterns = [];
for (const pkg of pkgsICareAbout) {
  copyPatterns.push({
    from: `${path.dirname(require.resolve(`${pkg}/package.json`))}/**/public/**/*`,
    noErrorOnMissing: true,
    to({ absoluteFilename }: { absoluteFilename: string }) {
      const regex = new RegExp('(public(?:\\\\|\/))(.*)');
      return regex.exec(absoluteFilename)![2];
    },
  })
}
```

But my main issue with this is the glob pattern's dont work.

For example, heres a snippet of what my node_modules looks like with the symlinks: 
image
image
The assets I want from this package are found in ui-framework/lib (this is what would show up in a regular node_module install), but that ui-framework pkg exists in my monorepo, so the entire src dir shows up instead of just the lib output dir as you'd see in a regular install.

But eh, i know for a fact that most assets i want will be in a `llib`, `build`, `esm`, or `cjs`, dir, so i can specify all those folders as well in my copyPatterns array. 

But this still doesn't work as its unable to correctly parse the glob pattern and won't copy a single thing over from the public subdir of those pkgs. I need to manually specify the file to copy over and it works, but its way too tedious to make a list of all the files in the public subdir to add to my copyPatterns array.

```ts
          {
            from: `${path.dirname(require.resolve(`@bentley/ui-framework/package.json`))}/lib/public/locales/en/UiFramework.json`,
            noErrorOnMissing: true,
            to({ absoluteFilename }) {
              console.log(absoluteFilename)
              return Promise.resolve(
                /(public\\)(.*)/.exec(absoluteFilename)[2]
              );
            },
          },
```

File: common/config/rush/browser-approved-packages.json
Comment (aruniverse) [inline common/config/rush/browser-approved-packages.json:939]: @calebmshafer does this look right to you? Should i revert this change?

File: .vscode/settings.json
Comment (calebmshafer) [inline .vscode/settings.json]: needs to be reverted before merge

File: common/config/rush/browser-approved-packages.json
Comment (calebmshafer) [reply common/config/rush/browser-approved-packages.json:939]: yea that looks like right since you removed the sass deps from the two test app.  Does CRA 4 bring in `sass` by default? We previously needed to call it out in ui-test-app due to the default being node-sass

**Review by calebmshafer — CHANGES_REQUESTED** [review]

File: common/config/rush/browser-approved-packages.json
Comment (aruniverse) [reply common/config/rush/browser-approved-packages.json:939]: yea cra should pull in sass directly now, so should be ok

File: test-apps/ui-test-app/.env
Comment (calebmshafer) [reply test-apps/ui-test-app/.env:1]: This is going to cause a handful of issues with our current workflow.  I do like the change but lets hold off since I want to ensure everyone doesn't accidentally now commit their .env with secrets in them.

I'm going to move all of these items back into the package.json directly rather than in a `.env`.

**Review by calebmshafer — APPROVED** [review]

File: test-apps/ui-test-app/.env.template
Comment (aruniverse) [inline test-apps/ui-test-app/.env.template:1]: @calebmshafer should i revert changes to this file if we're not using the `.env` anymore?

Comment (calebmshafer) [reply test-apps/ui-test-app/.env.template:1]: Ah I must have missed this when fixing everything.  Yes, lets revert it.

Comment (aruniverse) [reply test-apps/ui-test-app/.env.template]: reverted

**Pull request merged**
